src/Blocks/BlockPiston.cpp

```diff
@@ -154,171 +167,227 @@ bool cBlockPistonHandler::CanPushBlock(
+	else if (!a_BlocksPushed.insert(a_BlockPos).second || cBlockInfo::IsPistonBreakable(currBlock))
 	{
-		return true;  // Element exist already
+		//
+		// Element exist already
+		//
+		result = true;
 	}
-
-	if (currBlock == E_BLOCK_SLIME_BLOCK)
+	else
 	{
-		// Try to push the other directions
-		for (const auto & testDir : pushingDirs)
+		if (currBlock == E_BLOCK_SLIME_BLOCK)
 		{
-			if (!CanPushBlock(a_BlockPos + testDir, a_World, false, a_BlocksPushed, a_PushDir))
+			bool finished = false;
+			
+			//
+			// Try to push the other directions
+			//
+			for (const auto & testDir: pushingDirs)
 			{
-				// When it's not possible for a direction, then fail
-				return false;
+				if (finished)
+				{
+					break;
+				}
+				
+				result = CanPushBlock(a_BlockPos + testDir, a_World, false, a_BlocksPushed, a_PushDir);
+				
+				if (result == false)
+				{
+					//
+					// If we can't push the block, we're done.
+					//
+					finished = true;
+				}
 			}
+			
+			//
+			// All done here. DO PASS GO, do collect $100.
+			//
+		}
+		else
+		{
+			//
+			// Try to push the block in front of this block
+			//
+			result = CanPushBlock(a_BlockPos + a_PushDir, a_World, true, a_BlocksPushed, a_PushDir);
 		}
 	}
-
-	// Try to push the block in front of this block
-	return CanPushBlock(a_BlockPos + a_PushDir, a_World, true, a_BlocksPushed, a_PushDir);
+	
+	return result;
 }
 
 
 
 
-
 void cBlockPistonHandler::ExtendPiston(Vector3i a_BlockPos, cWorld * a_World)
 {
 	BLOCKTYPE pistonBlock;
 	NIBBLETYPE pistonMeta;
+	
 	a_World->GetBlockTypeMeta(a_BlockPos.x, a_BlockPos.y, a_BlockPos.z, pistonBlock, pistonMeta);
-
-	if (IsExtended(pistonMeta))
-	{
-		// Already extended, bail out
-		return;
-	}
-
-	Vector3i pushDir = MetadataToOffset(pistonMeta);
-
-	Vector3iSet blocksPushed;
-	if (!CanPushBlock(a_BlockPos + pushDir, a_World, true, blocksPushed, pushDir))
+	
+	if (!IsExtended(pistonMeta))
 	{
-		// Can't push anything, bail out
-		return;
+		Vector3i pushDir = MetadataToOffset(pistonMeta);
+		Vector3iSet blocksPushed;
+		
+		if (CanPushBlock(a_BlockPos + pushDir, a_World, true, blocksPushed, pushDir))
+		{
+			a_World->BroadcastBlockAction(a_BlockPos.x, a_BlockPos.y, a_BlockPos.z, 0, pistonMeta, pistonBlock);
+			a_World->BroadcastSoundEffect("tile.piston.out", a_BlockPos.x, a_BlockPos.y, a_BlockPos.z, 0.5f, 0.7f);
+			
+			PushBlocks(blocksPushed, a_World, pushDir);
+			
+			//
+			// Set the extension and the piston base correctly
+			//
+			Vector3i extensionPos = a_BlockPos + pushDir;
+			
+			a_World->SetBlock(
+				a_BlockPos.x, a_BlockPos.y, a_BlockPos.z,
+				pistonBlock, pistonMeta | 0x8);
+			
+			a_World->SetBlock(
+				extensionPos.x, extensionPos.y, extensionPos.z,
+				E_BLOCK_PISTON_EXTENSION, pistonMeta | (IsSticky(pistonBlock) ? 8 : 0)
+			);
+		}
+		else
+		{
+			//
+			// Can't push anything.
+			//
+			BLOCKTYPE blockAtPos = a_World->GetBlock(a_BlockPos + pushDir);
```

Comment:
This needs an ifdef guard, or moving the GetBlock call directly into the LOGD call.